### PR TITLE
Conjugate Gradient Lower Bound

### PR DESCRIPTION
**Pull request by awav**
CGLB [1] has a tighter bound and better hyperparameter estimation comparing to SGPR model. Implementation-wise the difference between SGPR and CGLB is an extra code for the conjugate gradient, which is quite simple. The original implementation can be found at [2].

[1]  https://arxiv.org/abs/2102.08314v1
[2] https://github.com/awav/CGLB

TODO list:

- [x] Add docstrings to key methods and classes
- [x] Notebook
- [x] Tests
    - [x] Conjugate gradient converges
    - [x] Set v = 0 and recover the quadratic term from SGPR
    - [x] Replace CG with another solver for K v = y

@DavidBurt2

### CODE REVIEW

File: gpflow/models/sgpr.py [hunk 1]
```diff
@@ -150,46 +151,95 @@ class SGPR(SGPRBase):
...
+        kuu = Kuu(iv, self.kernel, jitter=default_jitter())
         L = tf.linalg.cholesky(kuu)
-        sigma = tf.sqrt(self.likelihood.variance)
+        sigma = tf.sqrt(sigma_sq)
 
         # Compute intermediate matrices
         A = tf.linalg.triangular_solve(L, kuf, lower=True) / sigma
         AAT = tf.linalg.matmul(A, A, transpose_b=True)
-        B = AAT + tf.eye(num_inducing, dtype=default_float())
+        B = AAT + tf.eye(num_iv, dtype=default_float())
```
Comment (DavidBurt2) [inline gpflow/models/sgpr.py:178]: Using `set_diag` to avoid forming an identity matrix and adding explicitly would be preferable.

File: gpflow/models/sgpr.py [hunk 2]
```diff
@@ -150,46 +151,95 @@ class SGPR(SGPRBase):
...
+        x, y = self.data
+        num_data = to_default_float(tf.shape(x)[0])
+        output_dim = to_default_float(tf.shape(y)[1])
+        kdiag = self.kernel(x, full_cov=False)
+        sigma_sq = self.likelihood.variance
+
+        # tr(K - Q) / σ²
+        trace = tf.reduce_sum(kdiag) / sigma_sq - tf.reduce_sum(tf.linalg.diag_part(AAT))
+
+        # log(det(B))
```
Comment (DavidBurt2) [inline gpflow/models/sgpr.py:206]: this is really 0.5 * log(det(B)), right?

File: gpflow/models/cglb.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        return (0.5 * state.rz > max_error) and (state.i < max_steps)
+
+    def cg_step(state):
+        Ap = state.p @ K
+        denom = tf.reduce_sum(state.p * Ap, axis=-1)
+        gamma = state.rz / denom
+        v = state.v + gamma * state.p
+        i = state.i + 1
+        r = tf.cond(
+            state.i % restart_cg_step == restart_cg_step - 1,
```
Comment (DavidBurt2) [inline gpflow/models/cglb.py:138]: Is restarting necessary? I think we ended up not using it in practice. It might be good to check and remove it if not.

File: gpflow/models/cglb.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        # Correction term from Jensen's inequality
+        logtrace = num_data * tf.math.log(1 + trace / num_data)
+        return -output_dim * (logdet_b + 0.5 * logsigma_sq + 0.5 * logtrace)
+
+    def _quad_term(self, common) -> tf.Tensor:
+        x, y = self.data
+        err = y - self.mean_function(x)
+        num_data = to_default_float(tf.shape(err)[0])
+        sigma_sq = self.likelihood.variance
+        K = self.kernel.K(x) + sigma_sq * tf.eye(num_data, dtype=default_float())
```
Comment (DavidBurt2) [inline gpflow/models/cglb.py:55]: We should use `set_diag` to avoid forming the identity matrix here I think.

File: gpflow/models/sgpr.py [hunk 5]
```diff
@@ -150,46 +151,95 @@ class SGPR(SGPRBase):
...
+        """
+        common = self._common_calculation()
+        num_data = to_default_float(tf.shape(self.data[-1])[0])
+        output_dim = to_default_float(tf.shape(self.data[-1])[1])
+        const = -0.5 * num_data * output_dim * np.log(2 * np.pi)
+        logdet = self._logdet_term(common)
+        quad = self._quad_term(common)
+        return const + logdet + quad
 
     def predict_f(self, Xnew: InputData, full_cov=False, full_output_cov=False) -> MeanAndVariance:
```
Comment (DavidBurt2) [inline gpflow/models/sgpr.py:244]: `predict_f` could also make use of `_common_calculation` to remove redundant code I think.

File: gpflow/models/cglb.py (hunk 3 above)
Comment (awav) [reply gpflow/models/cglb.py:138]: Indeed, the restarting is never happened in practice. But I'm a bit hesitant to remove it. We know it can be effective as a preventative measure.

File: gpflow/models/sgpr.py (hunk 5 above)
Comment (awav) [reply gpflow/models/sgpr.py:243]: I would prefer to minimise changes in SGPR, and narrow the scope of this PR to CGLB only.

File: gpflow/models/sgpr.py [hunk 6]
```diff
@@ -150,46 +151,95 @@ class SGPR(SGPRBase):
...
 
+    CommonTensors = namedtuple("CommonTensors", ["A", "B", "LB", "AAT", "L"])
+
+
     def maximum_log_likelihood_objective(self) -> tf.Tensor:
         return self.elbo()
-
-    def elbo(self) -> tf.Tensor:
+    
+    def _common_calculation(self):
```
[the remark this answers is not in the document]
Comment (awav) [reply gpflow/models/sgpr.py:159]: https://github.com/GPflow/GPflow/pull/1706#discussion_r693400410

[the remark this answers is not in the document]
Comment (awav) [reply gpflow/models/sgpr.py:159]: Let me know if you think it is worth pushing for this change in this PR.

[the remark this answers is not in the document]
Comment (DavidBurt2) [reply gpflow/models/sgpr.py:159]: I guess it is fine for now and should perhaps be a separate PR. I do think it would simplify the code in a few places, and perhaps make it easier to do something in the future like cache values so that prediction is less expensive (i.e. avoiding redundant cholesky decompositions). But that is beyond the scope of this PR.

File: doc/source/notebooks/theory/cglb.pct.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+ls = np.linspace(0.01, 3, 100)
+losses_fn = lambda m: [loss_with_changed_parameter(m, m.kernel.lengthscales, l) for l in ls]
+gpr_obj = losses_fn(gpr)
+sgpr_obj = losses_fn(sgpr)
+cglb_obj = losses_fn(cglb)
+
+plt.plot(ls, gpr_obj, label="GPR")
+plt.plot(ls, sgpr_obj, label="SGPR")
+plt.plot(ls, cglb_obj, label="CGLB")
+plt.legend()
```
Comment (DavidBurt2) [inline doc/source/notebooks/theory/cglb.pct.py:56]: This should have axis labels. Maybe `lengthscale` and `Objective function`.

File: doc/source/notebooks/theory/cglb.pct.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+plt.show()
+
+
+# %% [markdown]
+# The interface of CGLB model is similar to SGPR, however, it has extra options related to the conjugate gradient which is involved in the computation of `yᵀK⁻¹y`.
```
Comment (DavidBurt2) [inline doc/source/notebooks/theory/cglb.pct.py:61]: Consider using latex math mode as opposed to quotes for math since this will be converted to markdown.

File: gpflow/models/cglb.py [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+        self._v = Parameter(tf.zeros((b, n), dtype=default_float()), trainable=v_grad_optimization)
+        self._cg_tolerance = cg_tolerance
+        self._max_cg_iters = max_cg_iters
+        self._restart_cg_iters = restart_cg_iters
+
+    @property
+    def aux_vec(self):
+        return self._v
+
+    def logdet_term(self, common):
```
Comment (vdutor) [inline gpflow/models/cglb.py:65]: what type is `common`?

File: gpflow/models/cglb.py [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+
+        kdiag = self.kernel(x, full_cov=False)
+        # t / σ²
+        trace = tf.reduce_sum(kdiag) / sigma_sq - tf.reduce_sum(tf.linalg.diag_part(AAT))
+        logdet_b = tf.reduce_sum(tf.math.log(tf.linalg.diag_part(LB)))
+        logsigma_sq = num_data * tf.math.log(sigma_sq)
+
+        # Correction term from Jensen's inequality
+        logtrace = num_data * tf.math.log(1 + trace / num_data)
+        return -output_dim * (logdet_b + 0.5 * logsigma_sq + 0.5 * logtrace)
```
Comment (vdutor) [inline gpflow/models/cglb.py:92]: shape information could be useful here for future maintenance.

File: gpflow/models/cglb.py [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+
+
+class CGLB(SGPR):
+    """
+    Conjugate Gradient Lower Bound. The key reference is
+
+    ::
+
+        @InProceedings{pmlr-v139-artemev21a,
+            title = {Tighter Bounds on the Log Marginal Likelihood of Gaussian Process Regression Using Conjugate Gradients},
```
Comment (vdutor) [inline gpflow/models/cglb.py:36]: can you reformat the text to fit the 100 character line-length?

File: gpflow/models/cglb.py [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+        .. math::
+            -0.5 * (rᵀQ⁻¹r + 2yᵀv - vᵀ K v ) <= -0.5 * yᵀK⁻¹y <= -0.5 * (2yᵀv - vᵀKv).
+
+        Equality holds if :math:`r=0`, i.e. :math:`v = K⁻¹y`.
+
+        If `self.aux_vec` is trainable, gradients are computed with respect to :math:`v` as well and :math:`v` can be optimized using
+        gradient based methods.
+
+        Otherwise, :math:`v` is updated with the method of conjugate gradients (CG).
+        CG is run until :math:`0.5 * rᵀQ⁻¹r <= ϵ`,  which ensures that the maximum bias due to this term is not more than :math:`ϵ`. The :math:`ϵ` is the CG tolerance.
```
Comment (vdutor) [inline gpflow/models/cglb.py:108]: can you break the line to fit the 100 character line-length?

File: gpflow/models/cglb.py [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+    """
+
+    def __init__(self, A, LB, sigma_sq):
+        self.A = A
+        self.LB = LB
+        self.sigma_sq = sigma_sq
+
+    def __call__(self, v):
+        """
+        Computes :math:`vQ^{-1}`. Note that this is implemented as multipication of a row vector on the right.
```
Comment (vdutor) [inline gpflow/models/cglb.py:279]: Looks like this method returns two thing, but the docstring says only one thing.

File: gpflow/models/cglb.py [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+        the initial position using the currect solution vector :math:`v`.
+        Can help avoid build up of numerical errors.
+    :return: Approximate solution to :math:`K v = b`.
+    """
+    CGState = namedtuple("CGState", ["i", "v", "r", "p", "rz"])
+
+    def stopping_criterion(state):
+        return (0.5 * state.rz > cg_tolerance) and (state.i < max_steps)
+
+    def cg_step(state):
```
Comment (vdutor) [inline gpflow/models/cglb.py:332]:
```suggestion
    def cg_step(state: "CGState"):
```

File: gpflow/models/cglb.py [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+    :param b: Vector we want to backsolve. Shape [B, N].
+    :param initial: Initial vector solution. Shape [N].
+    :param preconditioner: Preconditioner function.
+    :param cg_tolerance: Expected maximum error. This value is used as a
+        decision boundary against stopping criteria.
+    :param max_steps: Maximum number of CG iterations.
+    :param restart_cg_step: Restart step at which the CG resets the internal state to
+        the initial position using the currect solution vector :math:`v`.
+        Can help avoid build up of numerical errors.
+    :return: Approximate solution to :math:`K v = b`.
```
Comment (vdutor) [inline gpflow/models/cglb.py:325]: Is this correct? Does it return `b` or `v`?

File: gpflow/models/sgpr.py [hunk 16]
```diff
@@ -136,60 +137,114 @@ class SGPR(SGPRBase):
...
         # Compute intermediate matrices
         A = tf.linalg.triangular_solve(L, kuf, lower=True) / sigma
         AAT = tf.linalg.matmul(A, A, transpose_b=True)
-        B = AAT + tf.eye(num_inducing, dtype=default_float())
+        B = add_noise_cov(AAT, tf.cast(1.0, AAT.dtype))
         LB = tf.linalg.cholesky(B)
+
+        return self.CommonTensors(A, B, LB, AAT, L)
+
+    def logdet_term(self, common):
```
Comment (vdutor) [inline gpflow/models/sgpr.py:178]: what type is common?

File: gpflow/models/sgpr.py [hunk 17]
```diff
@@ -136,60 +137,114 @@ class SGPR(SGPRBase):
...
+        """
+        Bound from Jensen's Inequality:
+
+        log |K + σ²I| <= log |Q + σ²I| + N * log (1 + tr(K - Q)/(σ²N))
+
+        :param B: Matrix returned from common calc, unused
+        :param LB: Matrix returned from common calc
+        :param AAT: Matrix returned from common calc
```
Comment (vdutor) [inline gpflow/models/sgpr.py:186]: docstrings don't match input?

File: doc/source/notebooks/theory/cglb.pct.py [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+
+# %%
+v = tf.zeros_like(cglb.aux_vec)
+cglb.aux_vec.assign(v)
+
+# %%
+k = 100
+xnew = np.linspace(x.min() - 1.0, x.max() + 1.0, k).reshape(-1, 1)
+
+pred_no_tol = cglb.predict_y(xnew, cg_tolerance=None)
```
[the remark this answers is not in the document]
Comment (DavidBurt2) [reply doc/source/notebooks/theory/cglb.pct.py:106]: This means that no conjugate gradient will be run. 

The cached value of `self.aux_vec` will be used. If `self.aux_vec=0`, e.g. when the model is initialized, prediction will match `SGPR`. This is what is currently in the docstring for `predict_f`:

`cg_tolerance: float or None: If None, the cached value of :math:`v` is used. If float, conjugate gradient is run until :math:`rᵀQ⁻¹r < ϵ`.
`
Any suggestions on things we could do to make this clearer? Would it be useful to add this description to the notebook?

[the remark this answers is not in the document]
Comment (awav) [reply doc/source/notebooks/theory/cglb.pct.py:106]: @DavidBurt2, we should just re-state the same in the notebook.

File: doc/source/notebooks/theory/cglb.pct.py [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+
+fig, axes = plt.subplots(2, 1, figsize=(7, 5), sharex=True)
+fig.subplots_adjust(hspace=0.0)
+
+x, y = x.squeeze(), y.squeeze()
+xnew = xnew.squeeze()
+
+mu_no_tol = pred_no_tol[0].numpy().squeeze()
+std_no_tol = np.sqrt(pred_no_tol[1].numpy().squeeze())
+plot_prediction(axes[0], x, y, xnew, mu_no_tol, std_no_tol, "tab:blue", "CGLB")
```
[the remark this answers is not in the document]
Comment (awav) [reply doc/source/notebooks/theory/cglb.pct.py:117]:
```suggestion
plot_prediction(axes[0], x, y, xnew, mu_no_tol, std_no_tol, "tab:blue", "CGLB cached $v$")
```
